homeassistant/components/tado/__init__.py

```diff
@@ -38,91 +44,98 @@ def setup(hass, config):
     username = config[DOMAIN][CONF_USERNAME]
     password = config[DOMAIN][CONF_PASSWORD]
 
-    try:
-        tado = Tado(username, password)
-        tado.setDebugging(True)
-    except (RuntimeError, urllib.error.HTTPError):
-        _LOGGER.error("Unable to connect to mytado with username and password")
-        return False
+    tadoconnector = TadoConnector(hass, username, password)
+    hass.data[DOMAIN] = tadoconnector
 
-    hass.data[DATA_TADO] = TadoDataStore(tado)
+    # Do first update
+    tadoconnector.update()
 
+    # Load components
     for component in TADO_COMPONENTS:
-        load_platform(hass, component, DOMAIN, {}, config)
+        load_platform(
+            hass,
+            component,
+            DOMAIN,
+            {CONF_FALLBACK: config[DOMAIN][CONF_FALLBACK]},
+            config,
+        )
+
+    # Poll for updates in the background
+    hass.helpers.event.track_time_interval(
+        lambda now: tadoconnector.update(), SCAN_INTERVAL
+    )
 
     return True
 
 
-class TadoDataStore:
+class TadoConnector:
     """An object to store the Tado data."""
 
-    def __init__(self, tado):
-        """Initialize Tado data store."""
-        self.tado = tado
+    def __init__(self, hass, username, password):
+        """Initialize Tado Connector."""
+        self.hass = hass
 
-        self.sensors = {}
-        self.data = {}
+        try:
+            self.tado = Tado(username, password)
```

Comment:
We try to avoid side effects in init method. Move all side effects to a dedicated setup method.

Reply:
Thanks for the review! Could you explain what needs to be changed? Move this line to a separate function, and call that from the setup()?

Isn't it okay to instantiate something in the constructor?

Reply:
It's ok to instantiate in init but not if that has side effects. We're catching exceptions `RuntimeError` and `HTTPError` here so there must some side effects.

There might be more side effects in the calls below?

I suggest to move all this code to a new method `setup` on this instance.

Reply:
Those exceptions are for when your credentials are invalid for example.

I'm not totally sure what you mean with side effects. Is a HTTP call a side effect? Then this definitely has side effects :)

Reply:
Yes, I/O is always a side effect.

https://en.m.wikipedia.org/wiki/Side_effect_(computer_science)